File: service_python/nlp/token_tracker.py
from typing import Dict
import threading

class TokenTracker:
    def __init__(self):
        self.sessions: Dict[str, Dict[str, int]] = {}
        self.lock = threading.Lock()

    def add_tokens(self, input_tokens: int, output_tokens: int, session_id: str = "default"):
        """Add tokens to a session."""
        with self.lock:
            if session_id not in self.sessions:
                self.sessions[session_id] = {"input": 0, "output": 0, "total": 0}

            self.sessions[session_id]["input"] += input_tokens
            self.sessions[session_id]["output"] += output_tokens
            self.sessions[session_id]["total"] += input_tokens + output_tokens

    def get_total(self, session_id: str = "default") -> int:
        """Get total tokens for a session."""
        with self.lock:
            return self.sessions.get(session_id, {"total": 0})["total"]

    def get_breakdown(self, session_id: str = "default") -> Dict[str, int]:
        """Get token breakdown for a session."""
        with self.lock:
            return self.sessions.get(session_id, {"input": 0, "output": 0, "total": 0}).copy()

    def reset_session(self, session_id: str = "default"):
        """Reset token count for a session."""
        with self.lock:
            self.sessions[session_id] = {"input": 0, "output": 0, "total": 0}

    def list_sessions(self) -> Dict[str, Dict[str, int]]:
        """List all sessions and their token usage."""
        with self.lock:
            return self.sessions.copy()

File: extension/src/agent/tokenTracker.ts
import * as vscode from 'vscode';

/**
 * TokenTracker - Tracks token usage for conversation sessions
 * Accumulates input and output tokens across all LLM calls in a session
 */
export class TokenTracker {
    private totalTokens: number = 0;
    private webview?: vscode.Webview;

    constructor() {
        this.reset();
    }

    /**
     * Add tokens from an LLM call
     * @param inputTokens Number of input/prompt tokens
     * @param outputTokens Number of output/response tokens
     */
    addTokens(inputTokens: number, outputTokens: number): void {
        if (inputTokens < 0 || outputTokens < 0) {
            throw new Error('Token counts cannot be negative');
        }

        this.totalTokens += inputTokens + outputTokens;
        this.notifyWebview();
    }

    /**
     * Get the current total token count
     */
    getTotal(): number {
        return this.totalTokens;
    }

    /**
     * Reset the token counter for a new session
     */
    reset(): void {
        this.totalTokens = 0;
        this.notifyWebview();
    }

    /**
     * Set the webview reference for sending updates
     */
    setWebview(webview: vscode.Webview): void {
        this.webview = webview;
        // Send current total to initialize the UI
        this.notifyWebview();
    }

    /**
     * Send token update to the webview
     */
    private notifyWebview(): void {
        if (this.webview) {
            this.webview.postMessage({
                type: 'tokenUpdate',
                totalTokens: this.totalTokens
            });
        }
    }
}